NRLMMD-GEOIPS/geoips
# 74 classes for yaml based plugins

Note: initial 74 PR was accidentally merged, reverted, and this is the follow up PR.

Note all conversations are on the other PR that was merged / reverted.
https://github.com/NRLMMD-GEOIPS/geoips/pull/208

# Reviewer Checklist
https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md

# Related Issues
fixes NRLMMD-GEOIPS/geoips#74

For full 0 return, must also merge :
https://github.com/NRLMMD-GEOIPS/data_fusion/pull/12/files
https://github.com/NRLMMD-GEOIPS/geoips_clavrx/pull/9

# Testing Instructions

- Uninstall all geoips packages (e.g. geoips, data_fusion, geoips_clavrx)
- Reinstall geoips with `pip install .` (no `-e`).
- Run pytest (don't run test_full_install.sh yet, since we still need to be installed in editable mode with $GEOIPS_PACKAGES_DIR for the run_procflow tests to pass)
```
pip uninstall geoips
pip uninstall geoips_clavrx
pip uninstall data_fusion
pip uninstall recenter_tc
pip install $GEOIPS_PACKAGES_DIR/geoips
pip install $GEOIPS_PACKAGES_DIR/geoips_clavrx
pip install $GEOIPS_PACKAGES_DIR/data_fusion
pip install $GEOIPS_PACKAGES_DIR/recenter_tc

pytest tests/test_pytest
```

Then run this test from ipython:
```
from geoips import interfaces
for name in interfaces.__all__:
    inter = getattr(interfaces, name)
    print(name, len(inter.get_plugins()))
```
which should return something like this with no errors:
```
algorithms 10
colormaps 20
filename_formatters 9
interpolators 3
output_formatters 12
procflows 2
readers 28
title_formatters 3
products 26
product_defaults 75
feature_annotators 4
gridline_annotators 6
```

Now to double check that everything still works as it did before (ie, run from editable mode with all repos installed):
```
pip uninstall geoips
pip uninstall geoips_clavrx
pip uninstall data_fusion
pip uninstall recenter_tc
pip install -e $GEOIPS_PACKAGES_DIR/geoips
pip install -e $GEOIPS_PACKAGES_DIR/geoips_clavrx
pip install -e $GEOIPS_PACKAGES_DIR/data_fusion
pip install -e $GEOIPS_PACKAGES_DIR/recenter_tc
$GEOIPS_PACKAGES_DIR/geoips/tests/test_full_install.sh
```

# Summary


Version 1.10.0a11 (2023-04-30)
******************************

* Added classes to handle YAML-based plugins
* Added ability to discover YAML-based plugins from plugin packages

Major New Functionality
=======================

YAML-based plugin discovery without environment variables
---------------------------------------------------------

*From NRLMMD-GEOIPS/geoips#74: 2023-04-30, Add classes for yaml-based plugins*

* Added new load_all_yaml_plugins function in geoips_utils.py that is able to find
  yaml-based plugins
* Designed new entry point for registering entire plugin packages
* Added new BaseYamlInterface class
* Added new FeatureAnnotatorsInterface, GridlineAnnotatorsInterface,
  ProductDefualtsInterface, and ProductsInterface classes
* Completed functionality to merge Products with ProductDefaults plugins

::
    added: geoips/interfaces/module_based/__init__.py
    added: geoips/interfaces/yaml_based/__init__.py
    added: geoips/interfaces/yaml_based/feature_annotators.py
    added: geoips/interfaces/yaml_based/gridline_annotators.py
    added: geoips/interfaces/yaml_based/product_defaults.py
    added: geoips/interfaces/yaml_based/products.py
    added: geoips/plugins/__init__.py
    added: notes/cleanup.md
    added: tests/outputs/smap.unsectored.text_winds/smap-spd_rss_smap_surface_winds_20210926.0000.txt
    added: yaml-interface-notes.md
    deleted: geoips/dev/utils.py
    deleted: geoips/interfaces/boundaries.py
    deleted: tests/outputs/smap.unsectored.text_winds/smap-spd_rss_smap_surface_winds_20210926.0000.txt.gz
    modified: docs/source/releases/v1_10_0a8.rst
    modified: geoips/compare_outputs.py
    modified: geoips/geoips_utils.py
    modified: geoips/interfaces/__init__.py
    modified: geoips/interfaces/base.py
    modified: geoips/plugins/modules/interpolators/pyresample_wrappers/interp_gauss.py
    modified: geoips/plugins/modules/interpolators/pyresample_wrappers/interp_nearest.py
    modified: geoips/plugins/modules/interpolators/scipy_wrappers/interp_grid.py
    modified: geoips/plugins/modules/output_formatters/metadata_default.py
    modified: geoips/plugins/modules/output_formatters/metadata_tc.py
    modified: geoips/plugins/modules/output_formatters/netcdf_geoips.py
    modified: geoips/plugins/modules/procflows/config_based.py
    modified: geoips/plugins/modules/procflows/single_source.py
    modified: geoips/plugins/modules/readers/ahi_hsd.py
    modified: geoips/plugins/modules/sector_metadata_generators/bdeck_parser.py
    modified: geoips/plugins/yaml/product_defaults/pmw_37/19H.yaml
    modified: geoips/plugins/yaml/product_defaults/pmw_89/89H-Physical.yaml
    modified: geoips/plugins/yaml/products/ascatuhr.yaml
    modified: geoips/plugins/yaml/products/mimic.yaml
    modified: geoips/plugins/yaml/products/ssmis.yaml
    modified: geoips/plugins/yaml/sectors/static/goes17.yaml
    modified: geoips/schema/plugin_validator.py
    modified: geoips/schema/product_defaults/algorithm_colormap.yaml
    modified: geoips/schema/product_defaults/algorithm_interpolator_colormap.yaml
    modified: geoips/schema/product_defaults/interpolator_algorithm.yaml
    modified: geoips/schema/product_defaults/interpolator_algorithm_colormap.yaml
    modified: geoips/schema/products/bases/product.yaml
    modified: geoips/sector_utils/tc_tracks.py
    modified: interface_notes.md
    modified: pyproject.toml
    modified: tests/test_full_install.sh
    modified: yaml_interfaces.md
    renamed: geoips/interfaces/title_formatters.py	geoips/interfaces/module_based/title_formatters.py
    renamed: geoips/interfaces/interpolators.py	geoips/interfaces/module_based/interpolators.py
    renamed: geoips/interfaces/colormaps.py	geoips/interfaces/module_based/colormaps.py
    renamed: geoips/interfaces/readers.py	geoips/interfaces/module_based/readers.py
    renamed: geoips/interfaces/procflows.py	geoips/interfaces/module_based/procflows.py
    renamed: geoips/interfaces/algorithms.py	geoips/interfaces/module_based/algorithms.py
    renamed: geoips/interfaces/filename_formatters.py	geoips/interfaces/module_based/filename_formatters.py
    renamed: geoips/interfaces/output_formatters.py	geoips/interfaces/module_based/output_formatters.py

# Output
Package: geoips_full
Total run time: 2996 seconds
Number data types run: 37
Number data types failed: 0
